File: Networking/src/main/java/rpc/ClientRpcWorker.java
package rpc;

import dto.TripFilterDTO;
import model.domain.Agency;
import model.domain.Reservation;
import model.domain.Trip;
import model.utils.MyException;
import services.IObserver;
import services.IServices;
import services.ServiceException;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.Collection;

public class ClientRpcWorker implements Runnable, IObserver {
    private final IServices server;
    private final Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    private static final Response okResponse = new Response.Builder().type(ResponseType.OK).build();

    public ClientRpcWorker(IServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public void run() {
        while (connected) {
            try {
                Object request = input.readObject();
                Response response = handleRequest((Request)request);
                if (response != null)
                    sendResponse(response);
            } catch (IOException | ClassNotFoundException ex) {
                ex.printStackTrace();
            }
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException ex) {
            System.out.println("Error " + ex);
        }
    }

    private Response handleRequest(Request request) {
        if (request.type() == RequestType.LOGIN) return handleLogin(request);
        if (request.type() == RequestType.LOGOUT) return handleLogout(request);
        if (request.type() == RequestType.GET_AGENCIES) return handleGetAgencies(request);
        if (request.type() == RequestType.GET_TRIPS) return handleGetTrips(request);
        if (request.type() == RequestType.GET_RESERVATIONS) return handleGetReservations(request);
        if (request.type() == RequestType.SAVE_RESERVATION) return handleSaveReservation(request);

        return null;
    }

    private Response handleSaveReservation(Request request) {
        System.out.println("Save reservation request.." + request.type());
        Reservation reservation = (Reservation) request.data();
        try {
            server.saveReservation(reservation);
            return okResponse;
        } catch (ServiceException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGetReservations(Request request) {
        System.out.println("Get reservations request.." + request.type());
        try {
            Collection<Reservation> reservationsCol = server.getReservations();
            Reservation[] reservations = reservationsCol.toArray(new Reservation[0]);
            return new Response.Builder().type(ResponseType.GET_RESERVATIONS).data(reservations).build();
        } catch (ServiceException ex) {
            connected = false;
            return new Response.Builder().type(ResponseType.ERROR).data(ex.getMessage()).build();
        }
    }

    private Response handleGetTrips(Request request) {
        System.out.println("Get trips request.." + request.type());
        TripFilterDTO fDTO = (TripFilterDTO) request.data();
        try {
            Collection<Trip> tripsCol = server.getTrips(fDTO.destination(), fDTO.startTime(), fDTO.endTime());
            Trip[] trips = tripsCol.toArray(new Trip[0]);
            return new Response.Builder().type(ResponseType.GET_TRIPS).data(trips).build();
        } catch (ServiceException ex) {
            connected = false;
            return new Response.Builder().type(ResponseType.ERROR).data(ex.getMessage()).build();
        }
    }

    private Response handleGetAgencies(Request request) {
        System.out.println("Get agencies request.." + request.type());
        try {
            Collection<Agency> agenciesCol = server.getAgencies();
            Agency[] agencies = agenciesCol.toArray(new Agency[0]);
            return new Response.Builder().type(ResponseType.GET_AGENCIES).data(agencies).build();
        } catch (ServiceException ex) {
            connected = false;
            return new Response.Builder().type(ResponseType.ERROR).data(ex.getMessage()).build();
        }
    }

    private Response handleLogout(Request request) {
        System.out.println("Logout request.." + request.type());
        Agency agency = (Agency) request.data();
        try {
            server.logout(agency, this);
            connected = false;
            return okResponse;
        } catch (ServiceException ex) {
            return new Response.Builder().type(ResponseType.ERROR).data(ex.getMessage()).build();
        }
    }

    private Response handleLogin(Request request) {
        System.out.println("Login request.." + request.type());
        Agency agency = (Agency) request.data();
        try {
            server.login(agency, this);
            return okResponse;
        } catch (ServiceException ex) {
            connected = false;
            return new Response.Builder().type(ResponseType.ERROR).data(ex.getMessage()).build();
        }
    }

    private void sendResponse(Response response) throws IOException {
        System.out.println("Sending response " + response);
        output.writeObject(response);
        output.flush();
    }

    @Override
    public void ReservationSaved(Reservation reservation) throws ServiceException {
        Response response = new Response.Builder().type(ResponseType.NEW_RESERVATION).data(reservation).build();
        System.out.println("Reservation saved: " + reservation);
        try {
            sendResponse(response);
        } catch (IOException e) {
            throw new ServiceException("Reservation saved error: ", e);
        }
    }
}

File: Networking/rpc/ClientRpcWorker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Model;
using Networking.dto;
using Services;

namespace Networking.rpc
{
    public class ClientRpcWorker : IObserver
    {
        private readonly IServices _server;
        private TcpClient _connection;
        private NetworkStream _stream;
        private IFormatter _formatter;
        private volatile bool _connected;

        public ClientRpcWorker(IServices server, TcpClient connection)
        {
            _server = server;
            _connection = connection;
            try
            {
                _stream = _connection.GetStream();
                _formatter = new BinaryFormatter();
                _connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void Run()
        {
            while (_connected)
            {
                try
                {
                    object request = _formatter.Deserialize(_stream);
                    object response = HandleRequest((Request)request);
                    if (response != null)
                        SendResponse((Response)response);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.Message);
                }

                try
                {
                    Thread.Sleep(500);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                _stream.Close();
                _connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error: " + e);
            }
        }

        private void SendResponse(Response response)
        {
            Console.WriteLine("Sending response " + response);
            _formatter.Serialize(_stream, response);
            _stream.Flush();
        }

        private object HandleRequest(Request request)
        {
            if (request.Type == RequestType.LOGIN) return HandleLogin(request);
            if (request.Type == RequestType.LOGOUT) return HandleLogout(request);
            if (request.Type == RequestType.GET_AGENCIES) return HandleGetAgencies(request);
            if (request.Type == RequestType.GET_TRIPS) return HandleGetTrips(request);
            if (request.Type == RequestType.GET_RESERVATIONS) return HandleGetReservations(request);
            if (request.Type == RequestType.SAVE_RESERVATION) return HandleSaveReservation(request);
            return null;
        }

        private static readonly Response _okResponse = new Response.Builder().Type(ResponseType.OK).Build();
        private object HandleSaveReservation(Request request)
        {
            Console.WriteLine("Save reservation request.." + request.Type);
            Reservation reservation = (Reservation)request.Data;
            try
            {
                _server.SaveReservation(reservation);
                return _okResponse;
            }
            catch (ServiceException e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        private object HandleGetReservations(Request request)
        {
            Console.WriteLine("Get reservations request.." + request.Type);
            try
            {
                IEnumerable<Reservation> reservationsCol = _server.GetReservations();
                //Reservation[] reservations = reservationsCol.ToArray(new Reservation[0]);
                Reservation[] reservations = reservationsCol.ToArray();

                return new Response.Builder().Type(ResponseType.GET_RESERVATIONS).Data(reservations).Build();
            }
            catch (ServiceException ex)
            {
                _connected = false;
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.Message).Build();
            }
        }

        private object HandleGetTrips(Request request)
        {
            Console.WriteLine("Get trips request.." + request.Type);
            TripFilterDTO fDTO = (TripFilterDTO)request.Data;
            try
            {
                IEnumerable<Trip> tripsCol = _server.GetTrips(fDTO.Destination, fDTO.StartTime, fDTO.EndTime);
                //Trip[] trips = tripsCol.ToArray(new Trip[0]);
                Trip[] trips = tripsCol.ToArray();
                return new Response.Builder().Type(ResponseType.GET_TRIPS).Data(trips).Build();
            }
            catch (ServiceException ex)
            {
                _connected = false;
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.Message).Build();
            }
        }

        private object HandleGetAgencies(Request request)
        {
            Console.WriteLine("Get agencies request.." + request.Type);
            try
            {
                IEnumerable<Agency> agenciesCol = _server.GetAgencies();
                //Agency[] agencies = agenciesCol.ToArray(new Agency[0]);
                Agency[] agencies = agenciesCol.ToArray();
                return new Response.Builder().Type(ResponseType.GET_AGENCIES).Data(agencies).Build();
            }
            catch (ServiceException ex)
            {
                _connected = false;
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.Message).Build();
            }
        }

        private object HandleLogout(Request request)
        {
            Console.WriteLine("Logout request.." + request.Type);
            Agency agency = (Agency)request.Data;
            try
            {
                _server.Logout(agency, this);
                _connected = false;
                return _okResponse;
            }
            catch (ServiceException ex)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.Message).Build();
            }
        }

        private object HandleLogin(Request request)
        {
            Console.WriteLine("Login request.." + request.Type);
            Agency agency = (Agency)request.Data;
            try
            {
                _server.Login(agency, this);
                return _okResponse;
            }
            catch (ServiceException ex)
            {
                _connected = false;
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.Message).Build();
            }
        }

        public void ReservationSaved(Reservation reservation)
        {
            Response response = new Response.Builder().Type(ResponseType.NEW_RESERVATION).Data(reservation).Build();
            Console.WriteLine("Reservation saved: " + reservation);
            try
            {
                SendResponse(response);
            }
            catch (Exception e)
            {
                throw new ServiceException("Reservation saved error: ", e);
            }
        }
    }
}
